contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol

```diff
@@ -0,0 +1,1646 @@
+        }
+      }
+
+      // If there's a beneficiary, send the funds directly to the beneficiary. Otherwise send to the msg.sender.
+      _transferFrom(address(this), _beneficiary, netPayoutAmount);
+    }
+  }
+
+  /**
+    @notice
+    Takes a fee into the platform's project, which has an id of _FEE_BENEFICIARY_PROJECT_ID.
+
+    @param _projectId The ID of the project having fees taken from.
+    @param _fundingCycle The funding cycle during which the fee is being taken.
+    @param _amount The amount of the fee to take, as a floating point number with 18 decimals.
+    @param _beneficiary The address to mint the platforms tokens for.
+    @param _feeDiscount The amount of discount to apply to the fee, out of the MAX_FEE.
+
+    @return feeAmount The amount of the fee taken.
+  */
+  function _takeFeeFrom(
+    uint256 _projectId,
+    JBFundingCycle memory _fundingCycle,
+    uint256 _amount,
+    address _beneficiary,
+    uint256 _feeDiscount
+  ) internal returns (uint256 feeAmount) {
+    feeAmount = _feeAmount(_amount, fee, _feeDiscount);
+
+    if (_fundingCycle.shouldHoldFees()) {
+      // Store the held fee.
+      _heldFeesOf[_projectId].push(JBFee(_amount, uint32(fee), uint32(_feeDiscount), _beneficiary));
+
+      emit HoldFee(_projectId, _amount, fee, _feeDiscount, _beneficiary, msg.sender);
+    } else {
+      // Process the fee.
+      _processFee(feeAmount, _beneficiary, _projectId); // Take the fee.
+
+      emit ProcessFee(_projectId, feeAmount, false, _beneficiary, msg.sender);
+    }
+  }
+
+  /**
+    @notice
+    Process a fee of the specified amount.
+
+    @param _amount The fee amount, as a floating point number with 18 decimals.
+    @param _beneficiary The address to mint the platform's tokens for.
+    @param _from The project ID the fee is being paid from.
+  */
+  function _processFee(
+    uint256 _amount,
+    address _beneficiary,
+    uint256 _from
+  ) internal {
+    // Get the terminal for the protocol project.
+    IJBPaymentTerminal _terminal = directory.primaryTerminalOf(_FEE_BENEFICIARY_PROJECT_ID, token);
+
+    // If this terminal's token is ETH, send it in msg.value.
+    uint256 _payableValue = token == JBTokens.ETH ? _amount : 0;
+
+    // Send the projectId in the metadata.
+    bytes memory _projectMetadata = new bytes(32);
+    _projectMetadata = bytes(abi.encodePacked(_from));
+
+    // Trigger any inherited pre-transfer logic if funds will be transferred.
+    if (address(_terminal) != address(this))
+       _beforeTransferTo(address(_terminal), _amount);
+
+    try
+      // Send the fee.
+      _terminal.pay{value: _payableValue}(
+        _FEE_BENEFICIARY_PROJECT_ID,
+        _amount,
+        token,
+        _beneficiary,
+        0,
+        false,
+        '',
+        _projectMetadata
+      )
+    {} catch {
+      // Add fee amount back to project's balance.
+      store.recordAddedBalanceFor(_from, _amount);
```

Comment:
See [this comment](https://github.com/jbx-protocol/juice-contracts-v3/pull/17/files#r1103702332)